src/zarr/testing/store.py

```diff
@@ -149,6 +149,58 @@ async def test_read_only_store_raises(self, open_kwargs: dict[str, Any]) -> None
         ):
             await store.delete("foo")
 
+    async def test_with_read_only_store(self, open_kwargs: dict[str, Any]) -> None:
+        kwargs = {**open_kwargs, "read_only": True}
+        store = await self.store_cls.open(**kwargs)
+        assert store.read_only
+
+        # Test that you cannot write to a read-only store
+        with pytest.raises(
+            ValueError, match="store was opened in read-only mode and does not support writing"
+        ):
+            await store.set("foo", self.buffer_cls.from_bytes(b"bar"))
+
+        # Check if the store implements with_read_only
+        try:
+            writer = store.with_read_only(read_only=False)
+
```

Comment:
To avoid the big `try` block, I'd be tempted to put the except block here, then put the rest of the code for testing stores that implement `store.with_read_only` underneath in a dedented block.

Reply:
thanks, done in [`dff7034` (#3138)](https://github.com/zarr-developers/zarr-python/pull/3138/commits/dff7034c0a30a30ad9050da0a90b1cf53a531cb6)